src/twisted/internet/test/test_endpoints.py

```diff
@@ -2699,6 +2698,41 @@ def test_otherConnectionsCancelled(self):
         self.assertEqual([], self.mreactor.getDelayedCalls())
 
 
+class HostnameEndpointBindAddressTypes(unittest.TestCase):
+    """
+    Tests that HostnameEndpoint accepts all specified types for the
+    'bindAddress=' argument.
+    """
+
+    def setUp(self):
+        self.drr = deterministicResolvingReactor(MemoryReactor(), ["127.0.0.1"])
+
+    def test_bytes(self):
+        ba = b"1.2.3.4"
+        ep = endpoints.HostnameEndpoint(self.drr, b"example.com", 80, bindAddress=ba)
+        self.assertEqual(ep._bindAddress, ("1.2.3.4", 0))
+
+    def test_str(self):
+        ba = "1.2.3.4"
+        ep = endpoints.HostnameEndpoint(self.drr, b"example.com", 80, bindAddress=ba)
+        self.assertEqual(ep._bindAddress, ("1.2.3.4", 0))
+
+    def test_tuple_bytes(self):
+        ba = (b"1.2.3.4", 1234)
+        ep = endpoints.HostnameEndpoint(self.drr, b"example.com", 80, bindAddress=ba)
+        self.assertEqual(ep._bindAddress, ("1.2.3.4", 1234))
+
+    def test_tuple_str(self):
+        ba = ("1.2.3.4", 1234)
+        ep = endpoints.HostnameEndpoint(self.drr, b"example.com", 80, bindAddress=ba)
+        self.assertEqual(ep._bindAddress, ("1.2.3.4", 1234))
+
+    def test_none(self):
+        ba = None
+        ep = endpoints.HostnameEndpoint(self.drr, b"example.com", 80, bindAddress=ba)
+        self.assertEqual(ep._bindAddress, None)
+
+
```

Comment:
It would be nice to add a quick docstring to each of these explaining what it is trying to test, but, not a blocker.

Reply:
yeah, I figured the docstring on the class covered it well enough